src/sys/socket/mod.rs

```diff
@@ -86,32 +86,44 @@ pub enum SockProtocol {
     KextControl = libc::SYSPROTO_CONTROL,
 }
 
-libc_bitflags!{
-    /// Additional socket options
-    pub struct SockFlag: c_int {
-        /// Set non-blocking mode on the new socket
-        #[cfg(any(target_os = "android",
-                  target_os = "dragonfly",
-                  target_os = "freebsd",
-                  target_os = "linux",
-                  target_os = "netbsd",
-                  target_os = "openbsd"))]
-        SOCK_NONBLOCK;
-        /// Set close-on-exec on the new descriptor
-        #[cfg(any(target_os = "android",
-                  target_os = "dragonfly",
-                  target_os = "freebsd",
-                  target_os = "linux",
-                  target_os = "netbsd",
-                  target_os = "openbsd"))]
-        SOCK_CLOEXEC;
-        /// Return `EPIPE` instead of raising `SIGPIPE`
-        #[cfg(target_os = "netbsd")]
-        SOCK_NOSIGPIPE;
-        /// For domains `AF_INET(6)`, only allow `connect(2)`, `sendto(2)`, or `sendmsg(2)`
-        /// to the DNS port (typically 53)
-        #[cfg(target_os = "openbsd")]
-        SOCK_DNS;
+cfg_if! {
+    if #[cfg(any(target_os = "macos",
```

Comment:
Note here: `nix::sys::socket::socket` fn API requires the linux specific `SockFlag::SOCK_NONBLOCK` `SockFlag::SOCK_CLOEXEC` and which is not included in macOS' libc -- the work around is to provide a bitflag that mimics such support.

Reply:
Why can't this be fixed in `libc`? Do these constants not exist on OS X and iOS?

Reply:
This is fundamentally an implementation concern.

It's because the `nix::sys::socket::socket` fn is not only assigning a socket, but also setting flags on the socket in the same function to match parity with Linux (which is an improvement). The normal API is defined in socket(2) as:
```c
int socket(int domain, int type, int protocol);
```
linux diverts from POSIX and BSD by supplying a `SOCK_NONBLOCK` flag as is defined in linux man socket(2) as:
```
SOCK_NONBLOCK
Set the O_NONBLOCK file status flag on the new open
file description.  Using this flag saves extra calls
to fcntl(2) to achieve the same result.

SOCK_CLOEXEC
Set the close-on-exec (FD_CLOEXEC) flag on the new
file descriptor.  See the description of the
O_CLOEXEC flag in open(2) for reasons why this may be
useful.
```

In summery:
* `SOCK_NONBLOCK` and `SOCK_CLOEXEC` are in linux since 2.6.27
* Some BSDs have conformed to this new API and such have a define in `libc`.
* macOS is the exception and sticks to directly calling `fcntl(2)`
* `nix` takes the high road and also only calls `fcntl(2)` with a new fourth parameter on the function call

Because `nix` provides a new API that does not conform to POSIX or BSD (or directly Linux), this patch creates virtual definitions for macOS and iOS that are backed by their equivalent `fcntl` flags.

I think this is the best course of action for two reasons:

0. Redefining the function at this stage would break compatibility
0. Defining virtual definitions for `SOCK_NONBLOCK` and `SOCK_CLOEXEC` are only used inside of `nix` and only for the socket fn.

Reply:
For the sake of conversation, the alternative is:

Removing `pub struct SockFlag` and having the API user call `fcntl` directly.

Thus, these function defs need to be changed (breaking downstream code):
```rust
pub fn socket<T: Into<Option<SockProtocol>>>(domain: AddressFamily, ty: SockType, flags: SockFlag, protocol: T) -> Result<RawFd> {
```
```rust
/// Create a pair of connected sockets
///
/// [Further reading](http://pubs.opengroup.org/onlinepubs/9699919799/functions/socketpair.html)
pub fn socketpair<T: Into<Option<SockProtocol>>>(domain: AddressFamily, ty: SockType, protocol: T, flags: SockFlag) -> Result<(RawFd, RawFd)> {
```
```rust
/// Accept a connection on a socket
///
/// [Further reading](http://man7.org/linux/man-pages/man2/accept.2.html)
pub fn accept4(sockfd: RawFd, flags: SockFlag) -> Result<RawFd> {
```
```rust
fn accept4_polyfill(sockfd: RawFd, flags: SockFlag) -> Result<RawFd> {
```